### DISCUSSION TITLE
showing popup from viewmodel

### DISCUSSION THREAD

**Discussion by ashah7** [unoplatform/uno#17522] [Q&A] [answered]
Hi
I'm trying to show my popup box from my viewmodel using binding but it doesn't seem to work, can anyone help?

I set my variable in the viewmodel using ObservableProperty

```
  [ObservableProperty]
  private Boolean openpopupformessages;
```

I bind `Openpopupformessages `to `IsOpen `to my xaml Popup

```
    <Popup x:Name="StandardPopupforMessages" IsOpen="{Binding Openpopupformessages}">
      <Border BorderBrush="{StaticResource ApplicationForegroundThemeBrush}" 
           Background="{StaticResource ApplicationPageBackgroundThemeBrush}"
           BorderThickness="2" Width="480" Height="320">
        <StackPanel HorizontalAlignment="Center" VerticalAlignment="Center">
          <TextBlock><Run Text="{Binding Messagefromhub}" /></TextBlock>
          <Button Content="Close" CornerRadius="0" Background="Gray" Width="120" MinHeight="25" HorizontalAlignment="Center" Margin="0,15,0,0" Click="CloseMessagePopupClicked"  />
        </StackPanel>
      </Border>
    </Popup>
```

during my code execution I set it to `true`;

`Openpopupformessages = true; `

But the the popup does not show

If I set the `Openpopupformessages = true`'  (as a test) during the start of my `MainViewModel `it shows fine

**Comment by morning4coffe-dev** [comment]
Hey @ashah7, how do you update the `Openpopupformessages` in your code? I tested with: 
MainViewModel.cs: 
```csharp
    public ICommand Command { get; set; }

    [ObservableProperty]
    private Boolean openpopupformessages = false;

    public MainViewModel()
    {
        Command = new RelayCommand(ExecuteCommand);
    }

    private void ExecuteCommand()
    {
        Openpopupformessages = !Openpopupformessages; //you can set it to just true, it will only open the popup
    }
```

MainPage.xaml: 
```xml
<Button Command="{Binding Command}" Grid.Row="1" Content="open" />

<Popup x:Name="StandardPopupforMessages"
       IsOpen="{Binding Openpopupformessages}">
  <Border BorderBrush="{StaticResource ApplicationForegroundThemeBrush}"
          Background="{StaticResource ApplicationPageBackgroundThemeBrush}"
          BorderThickness="2"
          Width="480"
          Height="320">
    <StackPanel HorizontalAlignment="Center"
                VerticalAlignment="Center">
      <TextBlock><Run Text="test" /></TextBlock>
      <Button Content="Close"
              CornerRadius="0"
              Background="Gray"
              Width="120"
              MinHeight="25"
              HorizontalAlignment="Center"
              Margin="0,15,0,0"
              Command="{Binding Command}" />
    </StackPanel>
  </Border>
</Popup>
```
And it seems to be showing and closing well:

https://github.com/user-attachments/assets/0c9df8fd-8611-41fe-bed7-58457b454031

**Reply by ashah7** [reply]
thanks for this

The `Openpopupformessages` is updated when I receive a message from my SignalR server hub using the below code



```
  connection.On<string>("BroadCastMessageEstablishment", (Messagefromhub) =>
  {

      Openpopupformessages = true;
      Console.WriteLine(Messagefromhub);
      
  });
```


This might be difficult for you to test, but I can see the code firing when a message is received from my hub

I have also tried the below behind a timer that fires the below every 10 seconds 

     
```
   clickTimerBroadCast = new System.Timers.Timer();
        clickTimerBroadCast.Interval = 10000;
        clickTimerBroadCast.Start();
        clickTimerBroadCast.Elapsed += OnClickTimerBroadCast;
```



    private void OnClickTimerBroadCast(Object source, System.Timers.ElapsedEventArgs e)
    {
          Openpopupformessages = !Openpopupformessages;
    }


But even then its intermittent, sometimes the popup comes up and sometimes it does not.  I expect it to open and close every 10 seconds consistently.

**Reply by morning4coffe-dev** [reply, marked as answer]
Ah, I see! Thank you for specifying the details in your scenario. For this, I think you will need to enqueue the request with a dispatcher, as this might be running on a different thread. As an example, you can do something like this:
```csharp
var dispatcher = Window.Current.DispatcherQueue;
dispatcher .TryEnqueue(() =>
{
      Openpopupformessages = !Openpopupformessages;
});
```
Please let me know if this helps for your scenario as I didn't test it on my side. We can investigate further if not.

**Reply by ashah7** [reply]
excellent that looks like its working great with the `DispatcherQueue`

```
 connection.On<string>("BroadCastMessageEstablishment", (Messagefromhub) =>
 {
     var dispatcher = Window.Current.DispatcherQueue;
     dispatcher.TryEnqueue(() =>
     {
         Openpopupformessages = true;
     });
     Console.WriteLine(Messagefromhub);
 });
```

My `Messagefromhub `string doesnt seem to be showing on the Textbox binding, unsure why.  It does update consistently if I put it behind a Timer unlike the popupbox.  I'll open another discussion for that one
